src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -970,36 +1148,41 @@ public void RemoveFile(string fullPath)
                         throw new ArgumentException($"'{fullPath}' is not a source file of this project.");
                     }
 
-                    _documentPathsToDocumentIds.Remove(fullPath);
-
                     _project._documentFileChangeContext.StopWatchingFile(_project._fileWatchingTokens[documentId]);
                     _project._fileWatchingTokens.Remove(documentId);
 
-                    // There are two cases:
-                    // 
-                    // 1. This file is actually been pushed to the workspace, and we need to remove it (either
-                    //    as a part of the active batch or immediately)
-                    // 2. It hasn't been pushed yet, but is contained in _documentsAddedInBatch
-                    if (_documentAlreadyInWorkspace(_project._workspace.CurrentSolution, documentId))
+                    RemoveFileInternal(documentId, fullPath);
+                }
+            }
+
+            private void RemoveFileInternal(DocumentId documentId, string fullPath)
+            {
+                _documentPathsToDocumentIds.Remove(fullPath);
```

Comment:
it seems like this function expects to be taken while under the lock.  This definitely needs documentation.  And we should have an `Assert(Monitor.IsEntered...` here to make sure no one is violating this.